### PR TITLE
Load parent package config

### LINKED ISSUE #150196: Adapt flutter tools to handle pub workspaces:
With introduction of pub workspaces https://github.com/dart-lang/pub/issues/4127
[flutter.dev/go/pub-workspace](http://flutter.dev/go/pub-workspace)

One known issue is that .package_config is not always located relative to the apps pubspec.yaml, but can be located in a parent directory:

https://github.com/flutter/flutter/blob/7b6d667701e38c1d9d700e2bcc29bc7a06598c59/packages/flutter_tools/lib/src/dart/pub.dart#L653

### PR DESCRIPTION
**Pull request by sigurdm**
Fixes #150196

### CODE REVIEW

**Comment by andrewkolos** [comment]
Looks like this test is failing: https://github.com/flutter/flutter/blob/0a226c328e2941f86002250edc55e03400158d0d/packages/flutter_tools/test/commands.shard/permeable/packages_test.dart#L320

File: packages/flutter_tools/lib/src/dart/package_map.dart [hunk 1]
```diff
@@ -16,6 +16,44 @@ Future<PackageConfig> currentPackageConfig() async {
...
+    final File candidatePackagesFile = fileSystem.file(fileSystem.path.join(candidateDir, '.packages'));
+    if (fileSystem.file(candidatePackagesFile).existsSync()) {
+      return candidatePackagesFile;
+    }
+    final String nextDir = fileSystem.path.dirname(candidateDir);
+    if (nextDir == candidateDir) {
+      return null;
+    }
+    candidateDir = nextDir;
+  }
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/dart/package_map.dart:45]: Nit: I think this would be slightly more readable.

```suggestion
    final String parentDir = fileSystem.path.dirname(candidateDir);
    
    if (fileSystem.identicalSync(parentDir, candidateDir)) {
      return null;
    }
    candidateDir = parentDir;
```

File: packages/flutter_tools/lib/src/dart/package_map.dart [hunk 2]
```diff
@@ -16,6 +16,44 @@ Future<PackageConfig> currentPackageConfig() async {
...
+      fileSystem.path.join(candidateDir, '.dart_tool', 'package_config.json'),
+    );
+    if (fileSystem.file(candidatePackageConfigFile).existsSync()) {
+      return candidatePackageConfigFile;
+    }
+    // TODO(sigurdm): we should not need to check this file, it is obsolete.
+    final File candidatePackagesFile = fileSystem.file(fileSystem.path.join(candidateDir, '.packages'));
+    if (fileSystem.file(candidatePackagesFile).existsSync()) {
+      return candidatePackagesFile;
+    }
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/dart/package_map.dart:39]: If .packages is obsolete? Can we just remove this? If not, consider filing a follow-up issue and linking it here (with P3 and tech-debt labels).

File: packages/flutter_tools/lib/src/dart/package_map.dart [hunk 3]
```diff
@@ -16,6 +16,44 @@ Future<PackageConfig> currentPackageConfig() async {
...
+    final String nextDir = fileSystem.path.dirname(candidateDir);
+    if (nextDir == candidateDir) {
+      return null;
+    }
+    candidateDir = nextDir;
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/dart/package_map.dart:44]: Nit: I think this would be slightly more readable.
```suggestion
    final String parentDir = fileSystem.path.dirname(candidateDir);
    if (fileSystem.identicalSync(parentDir, candidateDir)) {
      return null;
    }
    candidateDir = parentDir;
```

**Comment by andrewkolos** [comment]
Is this ready for review? I see failing checks, but I don't know if these are flakes

**Comment by sigurdm** [comment]
I think all flutter tests are passing now. And the PR can be reviewed - but we still need a G3 hotfix.

Will work on that tomorrow.

**Comment by sigurdm** [comment]
And now it seems we're all green!

**Comment by sigurdm** [comment]
PTAL

**Comment by mit-mit** [comment]
@sigurdm looks like you need to update/rebase this?

File: packages/flutter_tools/lib/src/build_system/build_system.dart [hunk 4]
```diff
@@ -371,9 +372,11 @@ class Environment {
 
     final Directory rootBuildDir = buildDir ?? projectDir.childDirectory('build');
     final Directory buildDirectory = rootBuildDir.childDirectory(buildPrefix);
+    // final File packageConfigFile = findPackageConfigFileOrDefault(projectDir);
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/build_system/build_system.dart:375]:
```suggestion
```

File: packages/flutter_tools/lib/src/build_system/source.dart [hunk 5]
```diff
@@ -99,6 +99,12 @@ class SourceVisitor implements ResolvedFiles {
         // flutter root will not contain a symbolic link.
         Environment.kFlutterRootDirectory => environment.flutterRootDir.absolute.path,
         Environment.kProjectDirectory     => environment.projectDir.resolveSymbolicLinksSync(),
+        Environment.kWorkspaceDirectory   =>
+          environment.fileSystem.path.dirname(
+            environment.fileSystem.path.dirname(
+              environment.packageConfigPath,
+            )
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/build_system/source.dart:106]: turbo nit:
```suggestion
            ),
```

File: packages/flutter_tools/lib/src/build_system/build_system.dart [hunk 6]
```diff
@@ -457,6 +462,10 @@ class Environment {
   /// The [Source] value which is substituted with the path to [projectDir].
   static const String kProjectDirectory = '{PROJECT_DIR}';
 
+  /// The [Source] value which is substituted with the path to the parent of
+  /// [buildInfo.packageConfigPath].
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_system/build_system.dart:466]: nit, if you make this `[BuildInfo.packageConfigPath]`, editor jump to definition should work.

File: packages/flutter_tools/lib/src/build_system/build_system.dart [hunk 7]
```diff
@@ -457,6 +462,10 @@ class Environment {
...
+  static const String kWorkspaceDirectory = '{WORKSPACE_DIR}';
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_system/build_system.dart:467]: so, in practice, `{WORKSPACE_DIR}` == `.dart_tool/`?

File: packages/flutter_tools/lib/src/build_system/source.dart [hunk 8]
```diff
@@ -99,6 +99,12 @@ class SourceVisitor implements ResolvedFiles {
         // flutter root will not contain a symbolic link.
         Environment.kFlutterRootDirectory => environment.flutterRootDir.absolute.path,
         Environment.kProjectDirectory     => environment.projectDir.resolveSymbolicLinksSync(),
+        Environment.kWorkspaceDirectory   =>
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_system/source.dart:102]: oh, so the workspace dir is actually the PARENT of `.dart_tool/`?

File: packages/flutter_tools/lib/src/dart/package_map.dart [hunk 9]
```diff
@@ -16,6 +16,45 @@ Future<PackageConfig> currentPackageConfig() async {
   return loadPackageConfigUri(Isolate.packageConfigSync!);
 }
 
+/// Locates the `.dart_tool/package_config.json` relevant to [dir].
+///
+/// Searches [dir] and all parent directories.
+///
+/// Returns `null` if no package_config.json was found.
+File? findPackageConfigFile(Directory dir) {
+  final FileSystem fileSystem = dir.fileSystem;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/dart/package_map.dart:25]: nit, this can be a little cleaner by using more of the built-in APIs from package:file:

```dart
File? findPackageConfigFile(Directory dir) {
  final FileSystem fileSystem = dir.fileSystem;

  Directory candidateDir = fileSystem.directory(dir.path).absolute;
  while (true) {
    final File candidatePackageConfigFile = candidateDir
        .childDirectory('.dart_tool')
        .childFile('package_config.json');
    if (candidatePackageConfigFile.existsSync()) {
      return candidatePackageConfigFile;
    }
    // TODO(sigurdm): we should not need to check this file, it is obsolete,
    // https://github.com/flutter/flutter/issues/150908.
    final File candidatePackagesFile = candidateDir.childFile('.packages');
    if (candidatePackagesFile.existsSync()) {
      return candidatePackagesFile;
    }
    final Directory parentDir = candidateDir.parent;
    if (fileSystem.identicalSync(parentDir.path, candidateDir.path)) {
      return null;
    }
    candidateDir = parentDir;
  }
}
```

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 10]
```diff
@@ -694,7 +694,7 @@ class TestCommand extends FlutterCommand with DeviceBasedDevelopmentArtifacts {
   }) async {
     final AssetBundle assetBundle = AssetBundleFactory.instance.createBundle();
     final int build = await assetBundle.build(
-      packagesPath: '.packages',
+      packageConfigPath: '.dart_tool/package_config.json',
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/test.dart:697]: why do we hard-code here? do we not support workspaces in tests?

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 11]
```diff
@@ -258,50 +258,87 @@ class _DefaultPub implements Pub {
...
+    // Here we use pub's private helper file to locate the package_config.
+    // In pub workspaces pub will generate a `.dart_tool/pub/workspace_ref.json`
+    // inside each workspace-package that refers to the workspace root where
+    // .dart_tool/package_config.json is located.
+    //
+    // By checking for this file instead of iterating parent directories until
+    // finding .dart_tool/package_config.json we will not mistakenly find a
+    // package_config.json from outside the workspace.
+    //
+    // TODO(sigurdm): avoid relying on pubs private parts somehow?
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/dart/pub.dart:271]: nit I'd use the phrase "private implementation details" over "private parts"

File: packages/flutter_tools/lib/src/build_system/targets/native_assets.dart [hunk 12]
```diff
@@ -64,17 +64,15 @@ class NativeAssets extends Target {
...
       final TargetPlatform targetPlatform = getTargetPlatformForName(targetPlatformEnvironment);
       final Uri projectUri = environment.projectDir.uri;
-      final File packagesFile = fileSystem
-          .directory(projectUri)
-          .childDirectory('.dart_tool')
-          .childFile('package_config.json');
+
       final PackageConfig packageConfig = await loadPackageConfigWithLogging(
-        packagesFile,
+        environment.fileSystem.file(environment.packageConfigPath),
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_system/targets/native_assets.dart:69]: nit looks like `fileSystem` is already a local variable here that's equivalent to `environment.fileSystem`

File: packages/flutter_tools/lib/src/dart/package_map.dart [hunk 13]
```diff
@@ -16,6 +16,45 @@ Future<PackageConfig> currentPackageConfig() async {
...
+    }
+    candidateDir = parentDir;
+  }
+}
+
+/// Locates the `.dart_tool/package_config.json` relevant to [dir].
+///
+/// Like [findPackageConfigFile] but returns
+/// `$dir/.dart_tool/package_config.json` if no package config could be found.
+File findPackageConfigFileOrDefault(Directory dir) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/dart/package_map.dart:53]: It seems like we call this function a lot, is there any reason we can't cache it as a field in the `Environment` class?

File: packages/flutter_tools/lib/src/commands/assemble.dart [hunk 14]
```diff
@@ -217,10 +217,10 @@ class AssembleCommand extends FlutterCommand {
     return false;
   }
 
-  late final Environment _environment = _createEnvironment();
+  late final Future<Environment> _environment = _createEnvironment();
 
   /// The environmental configuration for a build invocation.
-  Environment _createEnvironment() {
+  Future<Environment> _createEnvironment() async {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/assemble.dart:223]: Why did this become async?

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 15]
```diff
@@ -258,50 +258,87 @@ class _DefaultPub implements Pub {
...
+      // This will incorrectly skip pub on the master branch if dependencies
+      // are being added/removed from the flutter framework packages, but this
+      // can be worked around by manually running pub.
+      if (checkUpToDate &&
+          pubLockFile.existsSync() &&
+          pubspecYaml.lastModifiedSync().isBefore(pubLockFile.lastModifiedSync()) &&
+          pubspecYaml.lastModifiedSync().isBefore(packageConfigFile.lastModifiedSync()) &&
+          lastVersion.existsSync() &&
+          lastVersion.readAsStringSync() == currentVersion.readAsStringSync()) {
       _logger.printTrace('Skipping pub get: version match.');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/dart/pub.dart:339]: I think this should be indented 2 spaces to the right.

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 16]
```diff
@@ -688,6 +729,10 @@ class _DefaultPub implements Pub {
     if (!project.manifest.generateSyntheticPackage) {
       return;
     }
+
+    if (!_fileSystem.path.equals(packageConfigFile.parent.parent.path, project.directory.path)) {
+      throwToolExit('`generate: true` is not supported in with workspaces.');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/dart/pub.dart:734]: nit: s/in with/within

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 17]
```diff
@@ -1750,21 +1768,22 @@ Run 'flutter -h' (or 'flutter <command> -h') for available flutter commands and
...
-        buildTargets: globals.buildTargets,
-      );
-
       await pub.get(
         context: PubContext.getVerifyContext(name),
         project: project,
         checkUpToDate: cachePubGet,
       );
 
+      await generateLocalizationsSyntheticPackage(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1781]: why did you move this after the pub.get? I would think that we would want to pub get after the synthetic package is created?

**Comment by christopherfujino** [comment]
Mostly nits, a few questions, but I think we can get this merged soon.

File: packages/flutter_tools/lib/src/build_system/source.dart (hunk 8 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/build_system/source.dart:102]: Yeah - I'll try to clarify!

File: packages/flutter_tools/lib/src/build_system/build_system.dart (hunk 6 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/build_system/build_system.dart:466]: Well - Should really be "grandparent" i guess. 
Tried to clarify

File: packages/flutter_tools/lib/src/build_system/build_system.dart (hunk 7 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/build_system/build_system.dart:467]: Well - Should really be "grandparent" i guess. 
Tried to clarify

File: packages/flutter_tools/lib/src/build_system/targets/native_assets.dart (hunk 12 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/build_system/targets/native_assets.dart:69]: Thanks!

File: packages/flutter_tools/lib/src/commands/assemble.dart (hunk 14 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/commands/assemble.dart:223]: I think at some point while I developed this it needed something async - seems that is no longer the case.

Made sync again.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 10 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/commands/test.dart:697]: I guess just to make existing tests continue working.
If we want to make many workspace tests it could totally be parametrized.

File: packages/flutter_tools/lib/src/dart/package_map.dart (hunk 13 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/dart/package_map.dart:53]: Yeah - this is a good point.

If we ensure that it is only called after pub get, then it should be fine.

I guess we are caching it in `BuildInfo` - everywhere should just read it from there.

Added a TODO

File: packages/flutter_tools/lib/src/dart/package_map.dart (hunk 9 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/dart/package_map.dart:25]: Nice!

File: packages/flutter_tools/lib/src/dart/pub.dart (hunk 16 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/dart/pub.dart:734]: Done

File: packages/flutter_tools/lib/src/runner/flutter_command.dart (hunk 17 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1781]: It depends on `environment` that now contains the package config path. That is not valid before pub get has run.
Also I don't see what pub has to use the synthetic package for...

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1781]: i trust you :)

**Review by christopherfujino — APPROVED** [review]
This LGTM if you add the TODO to [packages/flutter_tools/lib/src/commands/test.dart](https://github.com/flutter/flutter/pull/150850/files/2fa2ae2c525be708c52f3cfaad28a915553b6a01#diff-328d8a42e42b84152d577569aa043fdfa16977190165f1a11897f6f47d81c1a5)

**Review by andrewkolos — APPROVED** [review]
LGTM. Sorry for the silence.

File: packages/flutter_tools/lib/src/runner/flutter_command.dart (hunk 17 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1781]: I really don't have a good feel for the generateLocalizationsSyntheticPackage mechanism. But I don't believe it interacts with package resolution besides creating a package that is injected into package_config.json. So fingers crossed the order should not really matter.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 10 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/commands/test.dart:697]: Done

File: packages/flutter_tools/lib/src/runner/flutter_command.dart (hunk 17 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1781]: My naive understanding was that the project's package_config.json would need to have a reference to the synthetic package, and thus it would need to exist so that pub would find it and write it to package_config.json.

@jonahwilliams we might need your help here...

Comment (sigurdm) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1781]: pub itself will never write the synthetic package into package config. That is done by flutter_tool here:
https://github.com/flutter/flutter/blob/aaab2e145f20c87c27b39b7c4a382733a420e8fc/packages/flutter_tools/lib/src/dart/pub.dart#L751

That happens at the end of `flutter pub get`: https://github.com/flutter/flutter/blob/aaab2e145f20c87c27b39b7c4a382733a420e8fc/packages/flutter_tools/lib/src/dart/pub.dart#L369

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1781]: ohhhh, yeah, I forgot about that. :)
